File: src/backtest/metrics.py
"""Performance metrics calculation for backtesting.

Calculates Sharpe ratio, max drawdown, win rate, and other metrics.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional
import math


@dataclass
class PerformanceMetrics:
    """Performance metrics calculator for trading strategies."""

    max_drawdown: float = 0.0
    sharpe_ratio: float = 0.0
    sortino_ratio: float = 0.0
    calmar_ratio: float = 0.0
    win_rate: float = 0.0
    profit_factor: float = 0.0
    avg_trade: float = 0.0
    avg_win: float = 0.0
    avg_loss: float = 0.0
    largest_win: float = 0.0
    largest_loss: float = 0.0
    consecutive_wins: int = 0
    consecutive_losses: int = 0

    def reset(self) -> None:
        """Reset all metrics to initial state."""
        self.max_drawdown = 0.0
        self.sharpe_ratio = 0.0
        self.sortino_ratio = 0.0
        self.calmar_ratio = 0.0
        self.win_rate = 0.0
        self.profit_factor = 0.0
        self.avg_trade = 0.0
        self.avg_win = 0.0
        self.avg_loss = 0.0
        self.largest_win = 0.0
        self.largest_loss = 0.0
        self.consecutive_wins = 0
        self.consecutive_losses = 0

    def calculate_from_trades(
        self,
        trades: list[dict],
        equity_curve: list[dict],
        risk_free_rate: float = 0.02,
    ) -> None:
        """Calculate metrics from trade history and equity curve.

        Args:
            trades: List of trade dictionaries with 'pnl' key
            equity_curve: List of equity curve points [{timestamp, equity}]
            risk_free_rate: Annual risk-free rate for Sharpe calculation

        """
        if not trades:
            return

        # Calculate drawdown from equity curve
        self._calculate_drawdown(equity_curve)

        # Calculate trade statistics
        pnls = [t.get("pnl", 0) for t in trades]
        wins = [p for p in pnls if p > 0]
        losses = [p for p in pnls if p <= 0]

        self.win_rate = len(wins) / len(pnls) if pnls else 0

        if wins:
            self.avg_win = sum(wins) / len(wins)
            self.largest_win = max(wins)

        if losses:
            self.avg_loss = sum(losses) / len(losses)
            self.largest_loss = min(losses)  # Most negative

        self.avg_trade = sum(pnls) / len(pnls) if pnls else 0

        # Profit factor
        total_wins = sum(wins) if wins else 0
        total_losses = abs(sum(losses)) if losses else 0
        self.profit_factor = total_wins / total_losses if total_losses > 0 else 0

        # Derive annualization factor from equity curve timestamps
        trades_per_year = self._estimate_trades_per_year(trades)

        # Sharpe ratio
        self.sharpe_ratio = self._calculate_sharpe(pnls, risk_free_rate, trades_per_year)

        # Sortino ratio (downside deviation)
        self.sortino_ratio = self._calculate_sortino(pnls, risk_free_rate, trades_per_year)

        # Calmar ratio (return / max drawdown)
        if self.max_drawdown > 0:
            total_return = sum(pnls)
            self.calmar_ratio = total_return / self.max_drawdown

    def _estimate_trades_per_year(self, trades: list[dict]) -> float:
        """Estimate number of trades per year from close trade timestamps.

        Returns 252 as fallback if timestamps are unusable.
        """
        from datetime import datetime

        timestamps = []
        for t in trades:
            ts = t.get("timestamp")
            if ts is None:
                continue
            if isinstance(ts, datetime):
                timestamps.append(ts.timestamp())
            elif isinstance(ts, (int, float)):
                timestamps.append(float(ts))

        if len(timestamps) < 2:
            return 252.0

        first = min(timestamps)
        last = max(timestamps)
        span_seconds = last - first

        if span_seconds <= 0:
            return 252.0

        span_years = span_seconds / (365.25 * 24 * 3600)
        if span_years <= 0:
            return 252.0

        return len(timestamps) / span_years

    def _calculate_drawdown(self, equity_curve: list[dict]) -> None:
        """Calculate maximum drawdown from equity curve."""
        if not equity_curve:
            return

        peak = equity_curve[0]["equity"] if equity_curve else 0
        max_dd = 0.0

        for point in equity_curve:
            equity = point["equity"]
            if equity > peak:
                peak = equity
            drawdown = peak - equity
            if drawdown > max_dd:
                max_dd = drawdown

        self.max_drawdown = max_dd / peak * 100 if peak > 0 else 0.0

    def _calculate_sharpe(
        self, pnls: list[float], risk_free_rate: float, trades_per_year: float = 252.0
    ) -> float:
        """Calculate Sharpe ratio from PnL series.

        Args:
            pnls: Per-trade PnL values
            risk_free_rate: Annual risk-free rate
            trades_per_year: Estimated trades per year for annualization
        """
        if len(pnls) < 2:
            return 0.0

        import statistics

        avg_return = statistics.mean(pnls)
        std_return = statistics.stdev(pnls)

        if std_return == 0:
            return 0.0

        # Annualize using actual trade frequency
        sharpe = (
            (avg_return - risk_free_rate / trades_per_year)
            / std_return
            * math.sqrt(trades_per_year)
        )
        return sharpe

    def _calculate_sortino(
        self, pnls: list[float], risk_free_rate: float, trades_per_year: float = 252.0
    ) -> float:
        """Calculate Sortino ratio (uses downside deviation).

        Args:
            pnls: Per-trade PnL values
            risk_free_rate: Annual risk-free rate
            trades_per_year: Estimated trades per year for annualization
        """
        if len(pnls) < 2:
            return 0.0

        import statistics

        avg_return = statistics.mean(pnls)

        # Calculate downside deviations (only negative returns)
        negative_returns = [p for p in pnls if p < 0]
        if not negative_returns:
            return 0.0

        downside_std = statistics.stdev(negative_returns) if len(negative_returns) > 1 else 0

        if downside_std == 0:
            return 0.0

        # Annualize
        sortino = (
            (avg_return - risk_free_rate / trades_per_year)
            / downside_std
            * math.sqrt(trades_per_year)
        )
        return sortino

    def to_dict(self) -> dict:
        """Convert metrics to dictionary."""
        return {
            "max_drawdown": self.max_drawdown,
            "sharpe_ratio": self.sharpe_ratio,
            "sortino_ratio": self.sortino_ratio,
            "calmar_ratio": self.calmar_ratio,
            "win_rate": self.win_rate,
            "profit_factor": self.profit_factor,
            "avg_trade": self.avg_trade,
            "avg_win": self.avg_win,
            "avg_loss": self.avg_loss,
            "largest_win": self.largest_win,
            "largest_loss": self.largest_loss,
        }

    def __str__(self) -> str:
        """String representation of metrics."""
        return (
            f"Sharpe: {self.sharpe_ratio:.2f}, "
            f"Max DD: {self.max_drawdown:.2f}, "
            f"Win Rate: {self.win_rate:.1%}, "
            f"Profit Factor: {self.profit_factor:.2f}"
        )

File: src/backtest/metrics.rs
//! Performance metrics for backtest results.
//!
//! Pure functions where possible. Computes Sharpe ratio,
//! max drawdown, win rate, and other statistics.

/// Aggregate performance metrics.
#[derive(Debug, Clone)]
pub struct PerformanceMetrics {
    pub total_return_pct: f64,
    pub max_drawdown_pct: f64,
    pub max_drawdown_vs_initial: f64,
    pub sharpe_ratio: f64,
    pub win_rate: f64,
    pub total_trades: usize,
    pub winning_trades: usize,
    pub losing_trades: usize,
    pub avg_win: f64,
    pub avg_loss: f64,
    pub profit_factor: f64,
}

/// Compute Sharpe ratio from periodic returns. Pure function.
///
/// Uses risk-free rate = 0 (standard for crypto backtests).
/// Annualized = daily_sharpe * sqrt(365).
pub fn compute_sharpe(daily_returns: &[f64]) -> f64 {
    if daily_returns.len() < 2 {
        return 0.0;
    }
    let n = daily_returns.len() as f64;
    let mean = daily_returns.iter().sum::<f64>() / n;
    let variance = daily_returns.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / (n - 1.0);
    let std = variance.sqrt();
    if std == 0.0 {
        return 0.0;
    }
    let daily_sharpe = mean / std;
    // Annualize: trading days per year ≈ 365 for crypto
    daily_sharpe * 365.0_f64.sqrt()
}

/// Compute maximum drawdown from equity curve. Pure function.
///
/// Returns the maximum peak-to-trough drawdown as a fraction (0.0-1.0).
pub fn compute_max_drawdown(equity: &[f64]) -> f64 {
    if equity.len() < 2 {
        return 0.0;
    }
    let mut peak = equity[0];
    let mut max_dd = 0.0;
    for &value in equity.iter().skip(1) {
        if value > peak {
            peak = value;
        } else {
            let dd = (peak - value) / peak;
            if dd > max_dd {
                max_dd = dd;
            }
        }
    }
    max_dd
}

/// Compute maximum drawdown from initial capital. Pure function.
pub fn compute_max_drawdown_vs_initial(equity: &[f64], initial: f64) -> f64 {
    if initial <= 0.0 {
        return 0.0;
    }
    equity
        .iter()
        .map(|&e| (initial - e).max(0.0) / initial)
        .fold(0.0_f64, f64::max)
}

/// Compute win rate from trade P&Ls. Pure function.
pub fn compute_win_rate(pnls: &[f64]) -> f64 {
    if pnls.is_empty() {
        return 0.0;
    }
    let wins = pnls.iter().filter(|&&p| p > 0.0).count();
    wins as f64 / pnls.len() as f64
}

/// Compute profit factor (gross wins / gross losses). Pure function.
pub fn compute_profit_factor(pnls: &[f64]) -> f64 {
    let gross_wins: f64 = pnls.iter().filter(|&&p| p > 0.0).sum();
    let gross_losses: f64 = pnls.iter().filter(|&&p| p < 0.0).sum::<f64>().abs();
    if gross_losses == 0.0 {
        return if gross_wins > 0.0 { f64::INFINITY } else { 0.0 };
    }
    gross_wins / gross_losses
}

/// Compute daily returns from equity curve. Pure function.
pub fn compute_daily_returns(equity: &[f64]) -> Vec<f64> {
    if equity.len() < 2 {
        return Vec::new();
    }
    equity
        .windows(2)
        .map(|w| (w[1] - w[0]) / w[0])
        .collect()
}

/// Compute all metrics from trade records and equity curve.
pub fn compute_metrics(equity_curve: &[f64], closed_pnls: &[f64], initial_capital: f64) -> PerformanceMetrics {
    let total_return_pct = if initial_capital > 0.0 {
        let final_eq = equity_curve.last().copied().unwrap_or(initial_capital);
        ((final_eq - initial_capital) / initial_capital) * 100.0
    } else {
        0.0
    };

    let max_drawdown_pct = compute_max_drawdown(equity_curve) * 100.0;
    let max_drawdown_vs_initial = compute_max_drawdown_vs_initial(equity_curve, initial_capital) * 100.0;

    let daily_returns = compute_daily_returns(equity_curve);
    let sharpe_ratio = compute_sharpe(&daily_returns);

    let total_trades = closed_pnls.len();
    let winning_trades = closed_pnls.iter().filter(|&&p| p > 0.0).count();
    let losing_trades = closed_pnls.iter().filter(|&&p| p <= 0.0).count();
    let win_rate = compute_win_rate(closed_pnls) * 100.0;

    let winning_pnls: Vec<f64> = closed_pnls.iter().filter(|&&p| p > 0.0).copied().collect();
    let losing_pnls: Vec<f64> = closed_pnls.iter().filter(|&&p| p <= 0.0).copied().collect();

    let avg_win = if winning_trades > 0 {
        winning_pnls.iter().sum::<f64>() / winning_trades as f64
    } else {
        0.0
    };

    let avg_loss = if losing_trades > 0 {
        losing_pnls.iter().sum::<f64>() / losing_trades as f64
    } else {
        0.0
    };

    let profit_factor = compute_profit_factor(closed_pnls);

    PerformanceMetrics {
        total_return_pct,
        max_drawdown_pct,
        max_drawdown_vs_initial,
        sharpe_ratio,
        win_rate,
        total_trades,
        winning_trades,
        losing_trades,
        avg_win,
        avg_loss,
        profit_factor,
    }
}

// ─── Tests ──────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sharpe_constant_returns() {
        // Constant 1% daily return → infinite Sharpe (zero variance)
        let returns = vec![0.01, 0.01, 0.01, 0.01, 0.01];
        let sharpe = compute_sharpe(&returns);
        // Should be 0.0 (variance = 0)
        assert!((sharpe - 0.0).abs() < 0.001);
    }

    #[test]
    fn test_sharpe_positive() {
        let returns = vec![0.02, -0.01, 0.015, 0.005, -0.005, 0.01];
        let sharpe = compute_sharpe(&returns);
        assert!(sharpe > 0.0, "Sharpe should be positive for mostly positive returns");
    }

    #[test]
    fn test_max_drawdown() {
        let equity = vec![100.0, 110.0, 90.0, 95.0, 85.0, 100.0];
        let dd = compute_max_drawdown(&equity);
        // Peak = 110, trough = 85 → DD = (110-85)/110 ≈ 0.2273
        assert!((dd - 0.2273).abs() < 0.001, "Expected ~0.227, got {dd}");
    }

    #[test]
    fn test_max_drawdown_monotonic() {
        let equity = vec![100.0, 105.0, 110.0, 115.0];
        let dd = compute_max_drawdown(&equity);
        assert!((dd - 0.0).abs() < 0.001, "No drawdown expected");
    }

    #[test]
    fn test_win_rate() {
        let pnls = vec![100.0, -50.0, 200.0, -30.0, 50.0];
        let wr = compute_win_rate(&pnls);
        assert!((wr - 0.6).abs() < 0.001, "Expected 60% win rate, got {wr}");
    }

    #[test]
    fn test_win_rate_empty() {
        assert!((compute_win_rate(&[]) - 0.0).abs() < 0.001);
    }

    #[test]
    fn test_profit_factor() {
        let pnls = vec![100.0, -50.0, 200.0, -30.0];
        let pf = compute_profit_factor(&pnls);
        // 300 / 80 = 3.75
        assert!((pf - 3.75).abs() < 0.001, "Expected 3.75, got {pf}");
    }

    #[test]
    fn test_profit_factor_no_losses() {
        assert!(compute_profit_factor(&[100.0, 50.0]).is_infinite());
    }

    #[test]
    fn test_compute_metrics() {
        let equity = vec![10000.0, 10100.0, 10200.0, 10150.0, 10300.0];
        let pnls = vec![150.0, -50.0, 200.0];
        let metrics = compute_metrics(&equity, &pnls, 10000.0);
        assert!((metrics.total_return_pct - 3.0).abs() < 0.1, "Return ~3%");
        assert!((metrics.win_rate - 66.67).abs() < 1.0, "Win rate ~66.7%");
        assert_eq!(metrics.total_trades, 3);
        assert!(metrics.avg_win > 0.0);
        assert!(metrics.avg_loss < 0.0);
    }

    #[test]
    fn test_daily_returns() {
        let equity = vec![100.0, 105.0, 110.0];
        let returns = compute_daily_returns(&equity);
        assert_eq!(returns.len(), 2);
        assert!((returns[0] - 0.05).abs() < 0.001, "Expected 5%, got {}", returns[0]);
        assert!((returns[1] - 0.0476).abs() < 0.001, "Expected ~4.76%, got {}", returns[1]);
    }

    #[test]
    fn test_daily_returns_insufficient() {
        assert!(compute_daily_returns(&[100.0]).is_empty());
    }
}
